File: src/main/java/net/vanillaoutsider/betterdogs/ai/HuntWhenHurtGoal.java
// Copyright (C) 2026 Dasik (Rifaditya) | GNU GPLv3
// Verified against: Minecraft 26.2
package net.vanillaoutsider.betterdogs.ai;

import net.minecraft.world.entity.LivingEntity;
import net.minecraft.world.entity.ai.goal.Goal;
import net.minecraft.world.entity.animal.Animal;
import net.minecraft.world.entity.animal.wolf.Wolf;
import net.minecraft.world.level.Level;
import net.vanillaoutsider.betterdogs.util.WildHuntHelper;

import java.util.Comparator;
import java.util.EnumSet;
import java.util.List;

/**
 * Dedicated single-purpose AI goal for wild wolves to desperately hunt small prey only when hurt.
 * Wild wolves hunt when health falls below bd_wild_hunt_health_threshold (default 50%).
 * Upon defeating prey, restores +4.0 HP (2 hearts) sustenance healing.
 */
public class HuntWhenHurtGoal extends Goal {

    private final Wolf wolf;
    private LivingEntity targetPrey;

    public HuntWhenHurtGoal(Wolf wolf) {
        this.wolf = wolf;
        this.setFlags(EnumSet.of(Goal.Flag.TARGET));
    }

    @Override
    public boolean canUse() {
        if (!WildHuntHelper.shouldHuntPrey(this.wolf)) {
            return false;
        }

        Level level = this.wolf.level();
        if (level == null || level.isClientSide()) {
            return false;
        }

        List<Animal> potentialPrey = level.getEntitiesOfClass(
                Animal.class,
                this.wolf.getBoundingBox().inflate(16.0),
                WildHuntHelper::isPrey
        );

        this.targetPrey = potentialPrey.stream()
                .min(Comparator.comparingDouble(this.wolf::distanceToSqr))
                .orElse(null);

        return this.targetPrey != null;
    }

    @Override
    public void start() {
        if (this.targetPrey != null) {
            this.wolf.setTarget(this.targetPrey);
        }
    }

    @Override
    public boolean canContinueToUse() {
        LivingEntity currentTarget = this.wolf.getTarget();
        return WildHuntHelper.shouldContinueHunting(this.wolf, currentTarget);
    }

    @Override
    public void tick() {
        if (this.targetPrey != null && !this.targetPrey.isAlive()) {
            WildHuntHelper.applySustenanceHealing(this.wolf);
            this.targetPrey = null;
            this.wolf.setTarget(null);
        }
    }

    @Override
    public void stop() {
        this.targetPrey = null;
        this.wolf.setTarget(null);
    }
}

File: src/main/kotlin/net/vanillaoutsider/betterdogs/ai/HuntWhenHurtGoal.kt
package net.vanillaoutsider.betterdogs.ai

import net.minecraft.world.entity.ai.goal.Goal
import net.minecraft.world.entity.animal.Animal
import net.minecraft.world.entity.animal.wolf.Wolf
import java.util.EnumSet

/**
 * AI Goal for wild wolves to hunt prey only when hurt.
 * Replaces vanilla always-hunting behavior.
 */
class HuntWhenHurtGoal(
    private val wolf: Wolf
) : Goal() {
    
    companion object {
        private const val HEALTH_THRESHOLD = 0.5f // 50% health
        
        // Prey animal types by name (safer than class references)
        private val PREY_TYPES = setOf("sheep", "rabbit", "chicken")
    }
    
    init {
        flags = EnumSet.of(Flag.TARGET)
    }
    
    override fun canUse(): Boolean {
        // Only for wild (non-tamed) wolves
        if (wolf.isTame) return false
        
        // Only hunt when health is low
        val healthPercent = wolf.health / wolf.maxHealth
        if (healthPercent >= HEALTH_THRESHOLD) return false
        
        return true
    }
    
    override fun start() {
        // Find nearest prey by checking entity type name
        val nearestPrey = wolf.level().getEntitiesOfClass(
            Animal::class.java,
            wolf.boundingBox.inflate(16.0)
        ) { entity ->
            val typeName = entity.type.descriptionId.lowercase()
            PREY_TYPES.any { typeName.contains(it) }
        }.minByOrNull { wolf.distanceTo(it) }
        
        if (nearestPrey != null) {
            wolf.target = nearestPrey
        }
    }
    
    override fun canContinueToUse(): Boolean {
        if (wolf.isTame) return false
        
        // Stop hunting if we're healed enough
        val healthPercent = wolf.health / wolf.maxHealth
        if (healthPercent >= 0.8f) return false
        
        // Continue if target is alive
        val target = wolf.target ?: return false
        return target.isAlive
    }
    
    override fun stop() {
        wolf.target = null
    }
}
